### PR TITLE
[ASR] Adding Speech separation models to NeMo

### PR DESCRIPTION
**Pull request by krishnacpuvvada**
# What does this PR do ?
Adds speech separation model to NeMo.

**Collection**: ASR

# Changelog 
- High level script for speech separation `examples/speech_tasks/separation/speech_separation.py`
- Sample config file `examples/speech_tasks/separation/conf/sep_transformer.yaml`
- Separation model `nemo/collections/asr/models/ss_model.py`
- Pre-processor `AudioToFeaturesConvPreprocessor` in `nemo/collections/asr/modules/audio_preprocessing.py`
- Decoder `nemo/collections/asr/modules/ss_decoder.py`
- Loss `nemo/collections/asr/losses/ss_losses/si_snr.py`
- Dataset `nemo/collections/asr/data/audio_to_audio.py`
- Inference `EncDecSpeechSeparationModel.extract_sources()`  in `nemo/collections/asr/models/ss_model.py`

# Usage

```python
# load model
cfg = {'init_from_nemo_model': '<path_to_nemo_ckpt>'}
cfg = OmegaConf.create(cfg)
model_cfg = ASRModel.restore_from(restore_path=cfg.init_from_nemo_model, return_config=True)
sep_model = EncDecSpeechSeparationModel(cfg=model_cfg)
sep_model.maybe_init_from_pretrained_checkpoint(cfg, map_location='cpu')

# run inference
paths2audio_files=[<list_of_audio_mixtures_to_separate>]
sep_model.extract_sources(
    paths2audio_files = paths2audio_files,
    save_dir='<path_to_save_separated_sources>',
    orig_sr=16000,
    num_sources=2,
    batch_size=1,
    num_workers=1,
)
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/transformer_modules.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+from nemo.collections.asr.parts.submodules.multi_head_attention import (
+    MultiHeadAttention,
+    RelPositionMultiHeadAttention,
+)
+from nemo.collections.asr.parts.utils.activations import Swish
+
+__all__ = ['TransformerEncoderLayer']
+
+
+class TransformerEncoderLayer(torch.nn.Module):
```
Comment (jbalam-nv) [inline nemo/collections/asr/parts/submodules/transformer_modules.py:29]: @stevehuang52 Could you check if this transformer module is the same that you are using for SLU?

Comment (stevehuang52) [reply nemo/collections/asr/parts/submodules/transformer_modules.py:29]: > @stevehuang52 Could you check if this transformer module is the same that you are using for SLU?

@jbalam-nv This one is  mostly the same as the one in `nemo.collections.nlp.modules.common.transformer.transformer_encoder.TransformerEncoder`, but adds a FFN before the transformer. I also use TransformerDecoder and SequenceGenerator from `nemo.collections.nlp.modules.common.transformer` in SLU. I think we can move the whole transformer module to `nemo.collections.common` and avoid reimplementing similar stuffs.

File: nemo/collections/asr/modules/dual_path_model.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+        x2 = x[:, :, hop:].contiguous().view(B, F, -1, chunk_len)
+        x = torch.cat([x1, x2], dim=3).view(B, F, -1, chunk_len).transpose(2, 3)
+        return x.contiguous(), pad_len
+
+    def _padding(self, x, chunk_len, hop):
+        """
+        pad for whole number of chunks
+        """
+        B, F, N = x.shape
+        pad_len = chunk_len - (hop + N % chunk_len) % chunk_len
```
Comment (yzhang123) [inline nemo/collections/asr/modules/dual_path_model.py:312]: I'm not sure if i'm mistaken. 

But if c=4, h= 2, N=7, then pad_len should be 1 right? but this computes 3 for some reason

File: nemo/collections/asr/modules/dual_path_model.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        B, F, N = x.shape
+        x, pad_len = self._padding(x, chunk_len, hop)
+        x1 = x[:, :, :-hop].contiguous().view(B, F, -1, chunk_len)
+        x2 = x[:, :, hop:].contiguous().view(B, F, -1, chunk_len)
+        x = torch.cat([x1, x2], dim=3).view(B, F, -1, chunk_len).transpose(2, 3)
+        return x.contiguous(), pad_len
+
+    def _padding(self, x, chunk_len, hop):
+        """
+        pad for whole number of chunks
```
Comment (yzhang123) [inline nemo/collections/asr/modules/dual_path_model.py:309]: return object desc?

File: nemo/collections/asr/modules/dual_path_model.py [hunk 4]
```diff
@@ -0,0 +1,321 @@
...
+        Args:
+            x: torch.Tensor
+                [B, F, N]
+                [B, N ,L]
+            chunk_len: length of chunks
+            hop : hop size
+
+        Return:
+            output: torch.Tensor
+                [B, F, C, Nc]
```
Comment (yzhang123) [inline nemo/collections/asr/modules/dual_path_model.py:297]: should be F and C be swapped?

File: nemo/collections/asr/modules/dual_path_model.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+            pad: padding used 
+        """
+        B, F, N = x.shape
+        x, pad_len = self._padding(x, chunk_len, hop)
+        x1 = x[:, :, :-hop].contiguous().view(B, F, -1, chunk_len)
+        x2 = x[:, :, hop:].contiguous().view(B, F, -1, chunk_len)
+        x = torch.cat([x1, x2], dim=3).view(B, F, -1, chunk_len).transpose(2, 3)
```
Comment (yzhang123) [inline nemo/collections/asr/modules/dual_path_model.py:304]: love this nice trick!

File: nemo/collections/asr/modules/dual_path_model.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+        x = self.conv2d(x)
+        # [B, F*num_speakers, C, Nc]
+        B, _, C, Nc = x.shape
+
+        x = x.view(B * self.num_speakers, -1, C, Nc)
+        # [B*num_speakers, F, C, Nc]
+
+        x = self._overlap_add(x, pad_len, hop)
+        x = self.output(x) * self.output_gate(x)
```
Comment (yzhang123) [inline nemo/collections/asr/modules/dual_path_model.py:252]: is this part of paper?

File: nemo/collections/asr/modules/dual_path_model.py [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+        x = self.activation(x)
+
+        x = x.transpose(0, 1)
+        # [num_speakers, B, F, N]
+
+        return x
+
+    def _overlap_add(self, x, pad_len, hop):
+        """
+        Merge through overlap and add
```
Comment (yzhang123) [inline nemo/collections/asr/modules/dual_path_model.py:268]: add dimenstion of return tensor

File: nemo/collections/asr/modules/transformer_encoder.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+            Defaults to 0.0.
+    """
+
+    def __init__(
+        self,
+        n_layers,
+        d_model,
+        ff_expansion_factor=4,
+        self_attention_model='abs_pos',
+        n_heads=4,
```
Comment (yzhang123) [inline nemo/collections/asr/modules/transformer_encoder.py:69]: can we reuse Transformer models NeMo or refactor common parts?

**Review by yzhang123 — CHANGES_REQUESTED** [review]
Thanks @krishnacpuvvada , this is great! Could you also add a jenkins test? and in a separate PR update nemo documentation or add a tutorial?

File: nemo/collections/asr/parts/submodules/transformer_modules.py (hunk 1 above)
Comment (jbalam-nv) [reply nemo/collections/asr/parts/submodules/transformer_modules.py:29]: @krishnacpuvvada Steve has a PR to move transformer from nlp to common, could you please check if you could reuse some of that implementation?

**Review by jbalam-nv — CHANGES_REQUESTED** [review]
